### ISSUE TITLE
CNN Studied

### ISSUE THREAD

**Issue by JYKlabs** [JYKlabs/JYKlabs#1]
import os
import torch
import torch.nn as nn
import torch.optim as optim
import numpy as np
from torch.utils.data import Dataset, DataLoader, SubsetRandomSampler
from sklearn.model_selection import train_test_split  # 수정된 부분
import time
from openpyxl import Workbook
import shutil

def log(message):
    print(f"[{time.strftime('%Y-%m-%d %H:%M:%S')}] {message}")

class CachedAudioDataset(Dataset):
    def __init__(self, cache_dir, base_dir, input_prefix):
        self.cache_dir = cache_dir
        self.base_dir = base_dir
        self.input_prefix = input_prefix
        self.files = [f for f in os.listdir(cache_dir) if f.startswith("input_")]
        log(f"Dataset initialized with {len(self.files)} files from {cache_dir}")
        self.gain_map = {5: 0, 10: 1, 15: 2, 20: 3, 25: 4, 30: 5, 35: 6}
        self.peak_map = {i: idx for idx, i in enumerate(range(0, 101, 10))}
        self.gains = []
        self.peaks = []
        self._load_conditions()
    
    def _load_conditions(self):
        log("Loading file pairs and extracting conditions...")
        file_pairs = []
        for peak in range(0, 101, 10):
            folder = os.path.join(self.base_dir, f"Peak Reduction {peak}")
            if not os.path.exists(folder):
                continue
            input_files = [f for f in os.listdir(folder) if f.startswith(self.input_prefix) and f.endswith(".wav")]
            for input_file in input_files:
                suffix = input_file.replace(self.input_prefix, '').replace('.wav', '')
                output_file = f"outboardinput_{suffix}.wav"
                output_path = os.path.join(folder, output_file)
                if os.path.exists(output_path):
                    file_pairs.append((os.path.join(folder, input_file), output_path))
        
        log(f"Found {len(file_pairs)} valid file pairs")
        for idx, (input_path, _) in enumerate(file_pairs):
            parts = os.path.basename(input_path).split("_")
            gain = int([p for p in parts if p.startswith("gain")][0].replace("gain", ""))
            peak = int([p for p in parts if p.startswith("peak")][0].replace("peak", ""))
            self.gains.append(self.gain_map[gain])
            self.peaks.append(self.peak_map[peak])

    def __len__(self):
        return len(self.files)
    
    def __getitem__(self, idx):
        input_spec = np.load(os.path.join(self.cache_dir, f"input_{idx}.npy"))
        output_spec = np.load(os.path.join(self.cache_dir, f"output_{idx}.npy"))
        frames = np.load(os.path.join(self.cache_dir, f"frames_{idx}.npy"))
        if isinstance(frames, np.ndarray) and frames.size == 1:
            frames = frames.item()
        return (torch.from_numpy(input_spec).float(), torch.from_numpy(output_spec).float(), frames,
                self.gains[idx], self.peaks[idx])

class ConditionalAudioCNN(nn.Module):
    def __init__(self, input_height=1025, time_frames=657, gain_values=7, peak_values=11):
        super(ConditionalAudioCNN, self).__init__()
        self.input_height = input_height
        self.time_frames = time_frames
        self.gain_embedding = nn.Embedding(gain_values, 16)
        self.peak_embedding = nn.Embedding(peak_values, 16)
        self.conv1 = nn.Conv2d(1, 16, kernel_size=3, padding=1)
        self.conv2 = nn.Conv2d(16, 32, kernel_size=3, padding=1)
        self.pool = nn.MaxPool2d(2, 2)
        self.condition_fc = nn.Linear(32, 128)
        self.fc1 = nn.Linear(32 * (input_height // 4) * (time_frames // 4) + 128, 128)
        self.fc2 = nn.Linear(128, input_height * time_frames)
        self.relu = nn.ReLU()

    def forward(self, x, gain, peak):
        gain_embed = self.gain_embedding(gain)
        peak_embed = self.peak_embedding(peak)
        condition = torch.cat([gain_embed, peak_embed], dim=1)
        condition = self.condition_fc(condition)
        
        x = x.unsqueeze(1)
        x = self.pool(self.relu(self.conv1(x)))
        x = self.pool(self.relu(self.conv2(x)))
        x = x.view(x.size(0), -1)
        x = torch.cat([x, condition], dim=1)
        x = self.relu(self.fc1(x))
        x = self.fc2(x)
        x = x.view(x.size(0), self.input_height, self.time_frames)
        return x

def split_dataset(dataset, train_ratio=0.8, val_ratio=0.1):
    dataset_size = len(dataset)
    indices = list(range(dataset_size))
    train_idx, temp_idx = train_test_split(indices, train_size=train_ratio, random_state=42)
    val_idx, test_idx = train_test_split(temp_idx, train_size=val_ratio/(val_ratio + 0.1), random_state=42)
    return train_idx, val_idx, test_idx

def train_model(model, train_loader, val_loader, test_loader, device, num_epochs=30, model_name="cnn", save_dir="/Volumes/Mac Hub 3 for Dropbox/cnn_save"):
    criterion = nn.MSELoss()
    optimizer = optim.Adam(model.parameters(), lr=0.001)
    model_dir = os.path.join(save_dir, model_name)
    os.makedirs(model_dir, exist_ok=True)
    log(f"Created model directory: {model_dir}")
    
    start_time = time.time()
    log(f"Starting training for {model_name}")

    epoch_logs = []
    batch_logs = {f"Epoch {i+1}": [] for i in range(num_epochs)}
    total_batches = num_epochs * len(train_loader)
    batch_count = 0

    for epoch in range(num_epochs):
        model.train()
        train_loss = 0.0
        epoch_start_time = time.time()
        batch_losses = []
        log(f"Starting Epoch {epoch+1}/{num_epochs}")
        for i, (inputs, targets, frames, gains, peaks) in enumerate(train_loader):
            batch_count += 1
            progress = (batch_count / total_batches) * 100
            inputs, targets = inputs.to(device), targets.to(device)
            gains, peaks = gains.to(device), peaks.to(device)
            optimizer.zero_grad()
            outputs = model(inputs, gains, peaks)
            mask = torch.ones_like(outputs)
            for b in range(inputs.shape[0]):
                mask[b, :, frames[b]:] = 0
            loss = criterion(outputs * mask, targets * mask)
            loss.backward()
            optimizer.step()
            train_loss += loss.item()
            batch_losses.append(loss.item())
            print(f"{model_name} - Epoch [{epoch+1}/{num_epochs}], Batch [{i+1}/{len(train_loader)}], "
                  f"Loss: {loss.item():.4f}, Progress: {progress:.2f}%", end='\r')

        train_loss /= len(train_loader)
        batch_logs[f"Epoch {epoch+1}"] = batch_losses

        model.eval()
        val_loss = 0.0
        with torch.no_grad():
            for inputs, targets, frames, gains, peaks in val_loader:
                inputs, targets = inputs.to(device), targets.to(device)
                gains, peaks = gains.to(device), peaks.to(device)
                outputs = model(inputs, gains, peaks)
                mask = torch.ones_like(outputs)
                for b in range(inputs.shape[0]):
                    mask[b, :, frames[b]:] = 0
                val_loss += criterion(outputs * mask, targets * mask).item()
        val_loss /= len(val_loader)

        epoch_log = f"Epoch [{epoch+1}/{num_epochs}], Train Loss: {train_loss:.4f}, Val Loss: {val_loss:.4f}, Time: {(time.time() - epoch_start_time) // 60:.0f}m {(time.time() - epoch_start_time) % 60:.2f}s"
        epoch_logs.append(epoch_log)
        print(f"\n{model_name} - {epoch_log}")

        model_path = os.path.join(model_dir, f"{model_name}_epoch_{epoch+1}.pth")
        torch.save(model.state_dict(), model_path)
        log(f"Saved model: {model_path}")
        time.sleep(1)

    model.eval()
    test_loss = 0.0
    with torch.no_grad():
        for inputs, targets, frames, gains, peaks in test_loader:
            inputs, targets = inputs.to(device), targets.to(device)
            gains, peaks = gains.to(device), peaks.to(device)
            outputs = model(inputs, gains, peaks)
            mask = torch.ones_like(outputs)
            for b in range(inputs.shape[0]):
                mask[b, :, frames[b]:] = 0
            test_loss += criterion(outputs * mask, targets * mask).item()
    test_loss /= len(test_loader)
    log(f"{model_name} - Test Loss: {test_loss:.4f}")

    excel_file = os.path.join(save_dir, f"{model_name}_training_logs.xlsx")
    backup1_file = os.path.join(save_dir, f"{model_name}_training_logs_backup1.xlsx")
    backup2_file = os.path.join(save_dir, f"{model_name}_training_logs_backup2.xlsx")
    log(f"Saving training logs to {excel_file}")
    wb = Workbook()
    ws_epoch = wb.active
    ws_epoch.title = "Epoch Logs"
    ws_epoch.append(["Epoch Info"])
    for log_entry in epoch_logs:
        ws_epoch.append([log_entry])

    for epoch, losses in batch_logs.items():
        ws_batch = wb.create_sheet(title=epoch)
        ws_batch.append(["Batch", "Train Loss"])
        for i, loss in enumerate(losses, 1):
            ws_batch.append([i, loss])
    
    for attempt in range(5):
        try:
            wb.save(excel_file)
            shutil.copy(excel_file, backup1_file)
            shutil.copy(excel_file, backup2_file)
            if os.path.exists(excel_file) and os.path.exists(backup1_file) and os.path.exists(backup2_file):
                log(f"Successfully saved training logs to {excel_file}, backups to {backup1_file}, {backup2_file}")
                break
            else:
                raise FileNotFoundError("Files not created")
        except Exception as e:
            log(f"Attempt {attempt+1}/5 failed to save {excel_file}: {e}")
            time.sleep(2)
    else:
        raise RuntimeError(f"Failed to save {excel_file} after 5 attempts")

    total_time = time.time() - start_time
    log(f"{model_name} - Training finished")
    log(f"{model_name} - Total training time: {total_time // 60:.0f} minutes {total_time % 60:.2f} seconds")

if __name__ == "__main__":
    base_dir = "/Volumes/Mac Hub 3 for Dropbox/Dropbox/연구/Audio Hardware Modeling/Data Set/Test Signals Results/Linear_chirp"
    cache_dir_a = "/Volumes/Mac Hub 3 for Dropbox/Dropbox/연구/Audio Hardware Modeling/cache_clone_a"
    cache_dir_b = "/Volumes/Mac Hub 3 for Dropbox/Dropbox/연구/Audio Hardware Modeling/cache_clone_b"
    save_dir = "/Volumes/Mac Hub 3 for Dropbox/cnn_save"

    log("Starting dataset loading...")
    dataset_clone_a = CachedAudioDataset(cache_dir_a, base_dir, "Fixed_SA2A_input_")
    train_idx_a, val_idx_a, test_idx_a = split_dataset(dataset_clone_a)
    train_loader_a = DataLoader(dataset_clone_a, batch_size=6, sampler=SubsetRandomSampler(train_idx_a), num_workers=0, pin_memory=True)
    val_loader_a = DataLoader(dataset_clone_a, batch_size=6, sampler=SubsetRandomSampler(val_idx_a), num_workers=0, pin_memory=True)
    test_loader_a = DataLoader(dataset_clone_a, batch_size=6, sampler=SubsetRandomSampler(test_idx_a), num_workers=0, pin_memory=True)

    dataset_clone_b = CachedAudioDataset(cache_dir_b, base_dir, "directardinput_")
    train_idx_b, val_idx_b, test_idx_b = split_dataset(dataset_clone_b)
    train_loader_b = DataLoader(dataset_clone_b, batch_size=6, sampler=SubsetRandomSampler(train_idx_b), num_workers=0, pin_memory=True)
    val_loader_b = DataLoader(dataset_clone_b, batch_size=6, sampler=SubsetRandomSampler(val_idx_b), num_workers=0, pin_memory=True)
    test_loader_b = DataLoader(dataset_clone_b, batch_size=6, sampler=SubsetRandomSampler(test_idx_b), num_workers=0, pin_memory=True)

    device = torch.device("mps" if torch.backends.mps.is_available() else "cpu")
    log(f"Using device: {device}")

    num_epochs = 30

    log("Training CNN for Clone A...")
    model_a = ConditionalAudioCNN(input_height=1025, time_frames=657, gain_values=7, peak_values=11).to(device)
    train_model(model_a, train_loader_a, val_loader_a, test_loader_a, device, num_epochs=num_epochs, model_name="cnn_clone_a", save_dir=save_dir)

    log("Training CNN for Clone B...")
    model_b = ConditionalAudioCNN(input_height=1025, time_frames=657, gain_values=7, peak_values=11).to(device)
    train_model(model_b, train_loader_b, val_loader_b, test_loader_b, device, num_epochs=num_epochs, model_name="cnn_clone_b", save_dir=save_dir)
